sdk/webpubsub/azure-messaging-webpubsub-client/src/main/java/com/azure/messaging/webpubsub/client/WebPubSubAsyncClient.java

```diff
@@ -0,0 +1,1073 @@
+        logger.atVerbose().addKeyValue("code", closeReason.getCloseCode()).log("Session closed");
+
+        if (clientState.get() == WebPubSubClientState.STOPPED) {
+            return;
+        }
+
+        if (isStoppedByUser.compareAndSet(true, false)
+            || clientState.get() == WebPubSubClientState.STOPPING) {
+            // send DisconnectedEvent
+            tryEmitNext(disconnectedEventSink, new DisconnectedEvent(connectionId, null));
+
+            // stopped by user
+            handleClientStop();
+        } else if (closeReason.getCloseCode() == 1008) {
+            // do not send DisconnectedEvent
+            // server likely send the DisconnectedMessage before close
+
+            // VIOLATED_POLICY
+            handleClientStop();
+        } else {
+            if (!webPubSubProtocol.isReliable() || reconnectionToken == null || connectionId == null) {
+                clientState.changeState(WebPubSubClientState.DISCONNECTED);
+                // send DisconnectedEvent
+                tryEmitNext(disconnectedEventSink, new DisconnectedEvent(connectionId, null));
+
+                handleNoRecovery().subscribe(null, thr -> {
+                    logger.atWarning()
+                        .log("Failed to auto reconnect session: " + thr.getMessage());
+                });
+            } else {
+                handleRecovery().timeout(RECOVER_TIMEOUT, Mono.defer(() -> {
+                    // client should be RECOVERING, after timeout
+                    clientState.changeState(WebPubSubClientState.DISCONNECTED);
+                    // send DisconnectedEvent
+                    tryEmitNext(disconnectedEventSink, new DisconnectedEvent(connectionId, null));
+
+                    // fallback
+                    return handleNoRecovery();
+                })).subscribe(null, thr -> {
+                    logger.atWarning()
+                        .log("Failed to recover or reconnect session: " + thr.getMessage());
+                });
+            }
+        }
+    }
+
+    private void handleMessage(Object webPubSubMessage) {
+        if (logger.canLogAtLevel(LogLevel.VERBOSE)) {
+            try {
+                String json = JacksonAdapter.createDefaultSerializerAdapter()
+                    .serialize(webPubSubMessage, SerializerEncoding.JSON);
+                logger.atVerbose().addKeyValue("message", json).log("Received message");
+            } catch (IOException e) {
+                throw logger.logExceptionAsError(
+                    new UncheckedIOException("Failed to serialize received message for VERBOSE logging", e));
+            }
+        }
+
+        if (webPubSubMessage instanceof GroupDataMessage) {
+            GroupDataMessage groupDataMessage = (GroupDataMessage) webPubSubMessage;
+            tryEmitNext(groupMessageEventSink, new GroupMessageEvent(
+                groupDataMessage.getGroup(),
+                groupDataMessage.getData(),
+                groupDataMessage.getDataType(),
```

Comment:
We should tryUpdate sequenceId first and if `sequenceId != null` && sequenceId > the largest cached sequenceId, then Emit event. Otherwise, it's a duplicated message, we should drop it. The same for `ServerDataMessage`